lib/Sockets.js

```diff
@@ -0,0 +1,227 @@
+                   }
+               }
+
+               if (message.type == 'message') {
+                   controller.receiveMessage(bot, message);
+               } else if (message.type == 'hello' && !reconnected) {
+
+                   var uid = message.user;
+                   if (!uid) {
+                       uid = guid();
+                       bot.send({
+                           type: 'hello',
+                           user: uid,
+                       });
+                       controller.trigger(message.type,[bot, message]);
+                   } else {
+                       controller.trigger('welcome_back',[bot, message]);
+                   }
+
+               } else {
+                   controller.trigger(message.type,[bot, message]);
+               }
+
+
+           });
+
+         });
+
+         ws.on('close', function(err) {
+             console.log('CLOSED', err);
+             bot.connected = false;
+         });
+
+       });
+
+       var interval = setInterval(function ping() {
+         wss.clients.forEach(function each(ws) {
+           if(ws.isAlive === false){
+             console.log('!!!!!!!!!!!!!!!!!!!! ws terminated !!!!!!!!!!!!!!!!!!!!');
+             return ws.terminate()
+           }
+          //  if (ws.isAlive === false) return ws.terminate()
+           ws.isAlive = false;
+           ws.ping('', false, true);
+         });
+       }, 30000);
+     }
+
+
+     controller.defineBot(function(botkit, config) {
+       var bot = {
+           type: 'socket',
+           botkit: botkit,
+           config: config || {},
+           utterances: botkit.utterances,
+       };
+
+       bot.startConversation = function(message, cb) {
+           botkit.startConversation(this, message, cb);
+       };
+
+       bot.createConversation = function(message, cb) {
+           botkit.createConversation(this, message, cb);
+       };
+
+       bot.send = function(message, cb) {
+           if (bot.connected) {
+               console.log('SEND', message);
+               try {
+                   console.log('sending type ahead');
+                   bot.ws.send(JSON.stringify({typing:true}));
+                   setTimeout(function() {
+                       console.log('sending for reals');
+                       try {
+                         bot.ws.send(JSON.stringify(message));
+                       } catch(err) {
+                         console.error('ERROR SENDING', err);
+                         if (cb) return cb(err, message);
+                       }
+                       if (cb) cb(null, message);
+
+                   },1000);
+               } catch(err) {
+                   console.error('ERROR SENDING', err);
+               }
+           } else {
+               console.log('not connected. wait to resend');
+               setTimeout(function() {
+                   bot.send(message, cb);
+               }, 3000);
+           }
+       }
```

Comment:
Is there an interface that all chat integrations have to implement?
Facebook and Slackbot_worker have `startTyping`, `stopTyping`, `replyWithTyping` and `reply` methods.
Sockets `bot.send` method behaves like [replyWithTyping](https://github.com/howdyai/botkit/blob/07b7833869cf95d6aa714ba9b946e5633314b7dd/lib/Facebook.js#L173), but it is more primitive, because it types for 1 second when other libraries calculate how long the typing should last.